syntax = "proto3";

package business;

message BusinessRequest {
  int32 id = 1;
}

message BusinessByNameRequest {
  string name = 1;
}

message NewBusinessRequest {
  string businessid = 1;
  string name = 2;
  float rating = 3;
  int32 review_count = 4;
  string address = 5;
  string category = 6;
  string city = 7;
  string state = 8;
  string country = 9;
  string zip_code = 10;
  float latitude = 11;
  float longitude = 12;
  string phone = 13;
  string price = 14;
  string image_url = 15;
  string url = 16;
  float distance = 17;
  string business_hours = 18;
}

message BusinessResponse {
  int32 id = 1;
  string businessid = 2;
  string name = 3;
  float rating = 4;
  int32 review_count = 5;
  string address = 6;
  string category = 7;
  string city = 8;
  string state = 9;
  string country = 10;
  string zip_code = 11;
  float latitude = 12;
  float longitude = 13;
  string phone = 14;
  string price = 15;
  string image_url = 16;
  string url = 17;
  float distance = 18;
  string business_hours = 19;
}

message BusinessByLocationRequest {
  float latitude = 1;
  float longitude = 2;
  float radius = 3; // in kilometers
}

message CategoryRequest {
  string category = 1;
}

message RatingRequest {
  float min_rating = 1;
}

message BusinessByProximityRequest {
  float latitude = 1;
  float longitude = 2;
  int32 limit = 3; // Number of results
}

message TrendingRequest {}

message BusinessListResponse {
  repeated BusinessResponse businesses = 1;
}

service BusinessService {
  rpc GetBusiness (BusinessRequest) returns (BusinessResponse);
  rpc AddBusiness (NewBusinessRequest) returns (BusinessResponse);
  rpc GetBusinessByName (BusinessByNameRequest) returns (BusinessResponse);
  rpc GetBusinessByLocation (BusinessByLocationRequest) returns (BusinessListResponse);
  rpc GetBusinessByCategory (CategoryRequest) returns (BusinessListResponse);
  rpc GetBusinessByRating (RatingRequest) returns (BusinessListResponse);
  rpc GetBusinessByProximity (BusinessByProximityRequest) returns (BusinessListResponse);
  rpc GetTrendingBusinesses (TrendingRequest) returns (BusinessListResponse);
}
